colossalai/zero/sharded_model/sharded_grad.py

```diff
@@ -0,0 +1,122 @@
+import os
+from typing import List, Optional, Tuple
+from xml.dom.expatbuilder import parseFragment
+
+import torch
+import torch.distributed as dist
+from colossalai.context.parallel_mode import ParallelMode
+from colossalai.core import global_context as gpc
+from torch.distributed import ProcessGroup
+from torch.nn.parameter import Parameter
+
+from ._zero3_utils import (alloc_storage, assert_in_engine, free_storage,
+                           get_shard)
+
+
+class ShardedGradient:
+    def __init__(self,
+                 param: Parameter,
+                 process_group: Optional[ProcessGroup] = None,
+                 offload_config: Optional[dict] = None
+                 ) -> None:
+        self.param = param
+        self.process_group = process_group or gpc.get_group(ParallelMode.DATA)
+        self.shard_idx = self.process_group.rank()
+        self.num_shards = self.process_group.size()
+        self.offload_config = offload_config
+
+        self._cpu_offload = offload_config.get('device', None) == 'cpu' if offload_config else False
+
+        sharded_param, _ = get_shard(param.data, self.shard_idx, self.num_shards)
+
+        self._is_sharded = self.num_shards > 1
+        self._orig_size = param.data.size()
+        self._shard_size = sharded_param.size()
+        del sharded_param
+
+        self._saved_grad_shard: Optional[torch.Tensor] = None
+        self._saved_full_grad: Optional[torch.Tensor] = None
+        self._cpu_grad: Optional[torch.Tensor] = None
+
+        if self._cpu_offload:
+            # this buffer will be held and reused every iteration
+            self._cpu_grad = torch.zeros_like(param.data, device='cpu').pin_memory()
+
+    @torch.no_grad()
+    def setup(self) -> None:
+        """This function will be called pre-backward
+
+        :raises AssertionError: Raise if grad shape is wrong
+        """
+        if self.param.grad is not None:
+            if self.param.grad.device != self.param.data.device:
+                # TODO: offload?
+                raise RuntimeError(
+                    'grad and param are on different device, grad {self.param.grad.device} vs. param {self.param.data.device}')
+                self.param.grad = None
```

Comment:
Code is a bit redundant. move self.param.grad  = None out of if.
remove usless of branch.